### PR TITLE
Cache summaries in new documenable attribute

### LINKED ISSUE #497: Pydoctor is slow if there are many subclasses (caching for summary stan)
```
$ wget -O nodes.py https://sourceforge.net/p/docutils/code/HEAD/tree/tags/docutils-0.17.1/docutils/nodes.py?format=raw
$ time pydoctor nodes.py
...
pydoctor nodes.py  14.44s user 0.27s system 99% cpu 14.744 total
```

`nodes.py` contains 133 classes, generating documentation for it with pydoctor takes over 14 seconds on my machine.

It seems like Pydoctor is doing a bunch of work (parsing reST, resolving links in reST) for each subclass, e.g:

```python
class paragraph(General, TextElement): pass
class compound(General, Element): pass
class container(General, Element): pass
class bullet_list(Sequential, Element): pass
class enumerated_list(Sequential, Element): pass
class list_item(Part, Element): pass
class definition_list(Sequential, Element): pass
class definition_list_item(Part, Element): pass
class term(Part, TextElement): pass
class classifier(Part, TextElement): pass
class definition(Part, Element): pass
class field_list(Sequential, Element): pass
class field(Part, Element): pass
class field_name(Part, TextElement): pass
class field_body(Part, Element): pass
```

I think it would make sense to introduce some caching to avoid doing the same work multiple times.

### PR DESCRIPTION
**Pull request by tristanlatr**
Fixes #497

- Create `Documentable.parsed_summary` to cache summaries
- Create `stanutils.parsed_doc_from_stan(stan: Tag) -> 'ParsedDocstring'` to work with the code we have the generated stan, and cache it into a parsed docstring.

The performance benefits are quite big: 
- minus one minute of build time for `tox -e twisted-apidoc` (which includes pulling all Twisted code)
- minus 25 seconds for `tox -e numpy-apidoc`
- same for `tox -e python-igraph-apidocs`
- minus 2m for `tox -e cpython-summary`
- minus 3m for `tox -e cpython-apidocs`

### CODE REVIEW

File: pydoctor/epydoc2stan.py [hunk 1]
```diff
@@ -975,6 +975,63 @@ def ensure_parsed_docstring(obj: model.Documentable) -> Optional[model.Documenta
...
+        super().__init__(fields=[])
+        self._fromstan = stan
+    def has_body(self) -> bool:
+        return True
+    def to_stan(self, docstring_linker: Any) -> Tag:
+        return self._fromstan
+    def to_node(self) -> Any:
+        raise NotImplementedError()
+
+def parsed_doc_from_stan(stan: Tag) -> 'ParsedDocstring':
```
Comment (tristanlatr) [inline pydoctor/epydoc2stan.py:990]:
```suggestion
def parsed_doc_from_stan(stan: Tag) -> 'ParsedDocstring':
    """
    Encapsulate the given L{Tag} into a L{ParsedDocstring}.
    """
```

File: pydoctor/epydoc2stan.py [hunk 2]
```diff
@@ -975,6 +975,63 @@ def ensure_parsed_docstring(obj: model.Documentable) -> Optional[model.Documenta
...
+    return _ParsedDocFromStan(stan)
+
+def ensure_parsed_summary(obj: model.Documentable) -> Tuple[Optional[model.Documentable], ParsedDocstring]:
```
Comment (tristanlatr) [inline pydoctor/epydoc2stan.py:993]:
```suggestion
def ensure_parsed_summary(obj: model.Documentable) -> Tuple[Optional[model.Documentable], ParsedDocstring]:
    """
    Ensures that the C{parsed_summary} attribute of a documentable is set to it's final value. 
    Do not generate summary twice.
    """
```

File: pydoctor/epydoc2stan.py [hunk 3]
```diff
@@ -975,6 +975,70 @@ def ensure_parsed_docstring(obj: model.Documentable) -> Optional[model.Documenta
     else:
         return None
 
+
+class _ParsedDocFromStan(ParsedDocstring):
```
Comment (tristanlatr) [inline pydoctor/epydoc2stan.py:979]:
```suggestion
class _ParsedDocFromStan(ParsedDocstring):
    """
    A dummy L{ParsedDocstring}. 
    
    L{to_stan} method simply returns back what's given to L{_ParsedDocFromStan.__init__}. 
    """
```

File: pydoctor/epydoc2stan.py [hunk 4]
```diff
@@ -975,6 +975,70 @@ def ensure_parsed_docstring(obj: model.Documentable) -> Optional[model.Documenta
...
+    """
+    Encapsulate the given L{Tag} into a L{ParsedDocstring}.
+    """
+    return _ParsedDocFromStan(stan)
+
+def ensure_parsed_summary(obj: model.Documentable) -> Tuple[Optional[model.Documentable], ParsedDocstring]:
+    """
+    Ensures that the C{parsed_summary} attribute of a documentable is set to it's final value. 
+    Do not generate summary twice.
+    """
```
Comment (tristanlatr) [inline pydoctor/epydoc2stan.py:1000]:
```suggestion
    
    @returns: Tuple: C{source}, C{parsed docstring}
    """
```

File: pydoctor/epydoc2stan.py [hunk 5]
```diff
@@ -975,6 +975,77 @@ def ensure_parsed_docstring(obj: model.Documentable) -> Optional[model.Documenta
...
+
+def parsed_doc_from_stan(stan: Tag) -> 'ParsedDocstring':
+    """
+    Encapsulate the given L{Tag} into a L{ParsedDocstring}.
+    """
+    return _ParsedDocFromStan(stan)
+
+def ensure_parsed_summary(obj: model.Documentable) -> Tuple[Optional[model.Documentable], ParsedDocstring]:
+    """
+    Ensures that the C{parsed_summary} attribute of a documentable is set to it's final value.
```
Comment (tristanlatr) [inline pydoctor/epydoc2stan.py:1003]:
```suggestion
    Ensures that the L{model.Documentable.parsed_summary} attribute of a documentable is set to it's final value. 
```

File: pydoctor/epydoc2stan.py [hunk 6]
```diff
@@ -975,6 +975,77 @@ def ensure_parsed_docstring(obj: model.Documentable) -> Optional[model.Documenta
...
+    def to_node(self) -> Any:
+        raise NotImplementedError()
+
+def parsed_doc_from_stan(stan: Tag) -> 'ParsedDocstring':
+    """
+    Encapsulate the given L{Tag} into a L{ParsedDocstring}.
+    """
+    return _ParsedDocFromStan(stan)
+
+def ensure_parsed_summary(obj: model.Documentable) -> Tuple[Optional[model.Documentable], ParsedDocstring]:
```
Comment (adiroiban) [inline pydoctor/epydoc2stan.py:1001]: Maybe make it private and use a name that suggest this is a getter

I think the "ensure" part is not that relevant.

For the high-level function call, it's important to know that it extracts the summary from an `documentable`

What and how it does it and what it ensures, is low-level implementation.

```suggestion
def _get_parsed_summary(obj: model.Documentable) -> Tuple[Optional[model.Documentable], ParsedDocstring]:
```

**Review by adiroiban — APPROVED** [review]
Looks good.

It is an improvement and the  code changes are not that complicated.

Great work!

---------

I did some double checks. All good. As reported by you :)

For twisted apidoc I see that this PR takes 5m36s while in another PR it took 6m10s - https://github.com/twisted/pydoctor/runs/5645495029?check_suite_focus=true#step:6:1


While for cpython here it took 27m46s while in another PR 29m34s.

**Comment by tristanlatr** [comment]
Thanks @adiroiban 

Enabled auto-merge for this PR.

**Pull request merged**